### ISSUE TITLE
Proposal for new block storage interface

### ISSUE THREAD

**Issue by rvagg** [ipld/js-datastore-car#9]
This library attempted to adapt `interface-datastore` as an abstraction over CAR files. It's not the worst fit, but `interface-datastore` isn't designed for the CID:Binary parings that we use as a primitive. It's designed for complex string keys that reach into blocks, and we just don't need or want that. It also doesn't fit the `@ipld/block` so you need to wrap and unwrap everything. Even now just forcing string keys through the interface means lots of conversion to and from.

We need a new block storage abstraction. Here's a proposal that's not too far from `interface-datastore` but I think gives us what we need (in my rough TypeScript, pointers or direct edits on my TS are welcome!):

```ts
interface Store {
  close():Promise<void>,

  has(cid:CID):Promise<boolean>,
  get(cid:CID):Promise<Block>, // throws a NotFoundError
  put(block:Block):Promise<void>,
  delete(CID):Promise<boolean>,

  blocks():AsyncIterable<Block>,
  cids():AsyncIterable<CID>
}
```

`Car` would add two more methods:

```ts
interface Car extends Store {
  setRoots([CID]):Promise<void>,
  getRoots():Promise<CID[]>
}
```

`Car` is created from a factory method depending on what format data you have, but each has limitations on which `Store` methods it'll support. Those that are unsupported will throw an `UnsupportedError`.

```ts
Car.fromBuffer(Uint8Array):Car             // supports close(), has(), get(), blocks(), cids(), getRoots()
Car.fromFile(string|FD):Car                // supports close(), has(), get(), blocks(), cids(), getRoots()
Car.fromFileIndexed(string|FD):Car         // supports close(), has(), get(), blocks(), cids(), getRoots()
Car.fromStreamComplete(ReadableStream):Car // supports close(), has(), get(), blocks(), cids(), getRoots()
Car.fromStream(ReadableStream):Car         // supports close(), blocks(), cids(), getRoots() (can only call blocks() _or_ cids() once)
Car.toStream(WritableStream):Car           // supports close(), put(), setRoots()
```

(these are already implemented, with the old interface, but I'm thinking of changing their names, `read` -> `from` and `write` -> `to`). FYI:

* `fromBuffer()`, `fromFile()` and `fromStreamComplete()` all perform a full parse and keep everything in memory for you to access on demand.
* `fromStream()` is mainly good for a non-buffering dump of blocks - turn a stream into a `Block` iterator with `blocks()`.
* `fromFileIndexed()` performs a full parse but only buffers CIDs and the offsets of the relevant bytes so reads of individual blocks can seek to that position.

I wouldn't consider any new interface to be locked-in, we need to some more experimenting with these interfaces to make sure they're what we need. This library only serves a narrow range of use-cases, it's not a general purpose storage system, it's just using a storage system abstraction for convenience.

Can anyone muster strong enough opinions to comment before I plough ahead? @vmx @mikeal @gozala, anyone else?

**Comment by Gozala** [comment]
Thanks @rvagg for including me in this thread, this is really interesting for me because I have been thinking about unification of dag-service API with car myself (some relevant ramblings here https://github.com/ipfs/js-dag-service/issues/338). Here are some of my notes and feedback (I have also updated interface definitions to a valid TS)

- > `get(cid:CID):Promise<Block>, // throws a NotFoundError`

  I don't personally like "not found" style exceptions, as it leads to either:
  1. try catch blocks all over the code base.
  2. errors in production because try catch was not used.

  I would very much prefer following:

   ```ts
   interface Store {
     get(cid:CID):Promise<Block|void>
     get <T> (cid:CID, notFound:T):Promise<Block|T>
   }
   ```

   Because:

   1. It forces user to deal with not found blocks (although technically one could still assume `Promise<Block>` is there)
   2. No try catch is required for handling not found blocks.
   3. It is similar to built-in map.
   4. `get` override (isn't necessary, but) simplifies `foo = (await store.get(cid)) || bla` to `foo = await store.get(cid, bla)`.

- > `Car` is created from a factory method depending on what format data you have, but each has limitations on which `Store` methods it'll support. Those that are unsupported will throw an `UnsupportedError`.

     I really dislike these kind of abstractions. And would highly recommend going with with separate interfaces & classes for individual use cases even if they end up overlapping quite a bit. There are direct implications on typings (but that is not the reason I find them problematic just another singnal). E.g. if you had `StoreReader` / `StoreWriter` interfaces and `toStream(stream:WritableStream):StoreWriter` type checker would not allow calling `get` (or anything that's not part of `StoreWriter`) interface, furthermore layers above could be more specific what they need and avoid having to deal with runtime errors.

    I am also guessing that some store instances that are both `StoreWriter` and `StoreReader` could technically provide synchronous reads. But due to abstracting everything has to go through `await`. By separating `SyncStoreReader` and `AsyncStoreReader` interfaces sync one still complies with async but not the other way round.

   I'd be glad to do a deeper sync on this if this seems like a discussion worth having.  

- For the dag-service I do want something along the lines of `Car` store (I was thinking `BlockPack`)  interface to enable batched writes.

**Comment by rvagg** [comment]
```ts
  get <T> (cid:CID, notFound:T):Promise<Block|T>
```

I just threw up a little.

So this `NotFound` problem reminds me that I made a decision already on this in the HAMT abstraction, doing what you say - mirror what `Map` does and return `undefined` since we don't support storing `undefined`. That should make it feel more natural. So let's just go with a simple `get(cid:CID):Promise<Block|void>` (is the `|void` doing the "may be `null` or `undefined` here or something else? Should it be `|undefined` if we want to be strict about not-found returning `undefined` so there's no additional optionality in what it returns?).

How about this collection then?

```ts
interface Store {
  close():Promise<void>
}
```

```ts
interface StoreReader extends Store {
  has(cid:CID):Promise<boolean>,
  get(cid:CID):Promise<Block|undefined>
  getRoots():Promise<CID[]>
}
```

```ts
interface StoreWriter extends Store {
  put(block:Block):Promise<void>,
  delete(CID):Promise<boolean>,
  setRoots([CID]):Promise<void>,
}
```

```ts
interface StoreIterator extends Store {
  blocks():AsyncIterable<Block>,
  cids():AsyncIterable<CID>,
  getRoots():Promise<CID[]>
}
```

The `setRoots()` and `getRoots()` (that last one's duplicated in `StoreReader` and `StoreIterator`, maybe there's an elegant way to factor that out without even more interfaces?) are interesting here. They're particular to the CAR storage abstraction, but maybe they're not so particular? They're kind of like pins, they're really just graph roots, so maybe they're entirely reasonable to include in generic `Store*` interfaces and therefore we don't need particular `Car*` extensions?

Then you'd have:

```ts
Car.fromBuffer(Uint8Array):StoreReader
Car.fromFile(string|FD):StoreReader
Car.fromFileIndexed(string|FD):StoreReader
Car.fromStreamComplete(ReadableStream):StoreReader
Car.fromStream(ReadableStream):StoreIterator
Car.toStream(WritableStream):StoreWriter
```

**Comment by rvagg** [comment]
OK, so the lack of multiple inheritance makes it awkward

```ts
Car.createReader(Uint8Array|string|FD|ReadableStream):StoreReader
Car.createIndexedReader(string|FD):StoreReader
Car.createIterator(Uint8Array|string|FD|ReadableStream):StoreIterator
Car.createWriter(WritableStream):StoreWriter
```

`createIndexedReader()` is an awkward one there, it could be folded into an argument: `Car.createReader(Uint8Array|string|FD|ReadableStream, boolean)` where the bool is "indexed" but only works if you provide a filename or FD.

I'd also be tempted to just remove `string|FD` from `createReader()` and leave that to the user, like `createWriter()`, it's there because it allowed me to write a more efficient file reader than just using `createReadableStream()` but it's really not that big a deal and I could offload that burden. _Perhaps_ if you provide a `string|FD` to `createReader` then you'll get an indexed one, but that might be surprising behaviour to the user who just wants to read a simple CAR file. That's one reason why having a separate factory for the "indexed" case is nice.

Also I think maybe I can factor our `ReadableStream` and `WritableStream` and just accept and return `AsyncIterator<Uint8Array>`, I think that's safe enough to do these days. That would offer quite a bit more flexibility and simplify dependency trees in the browser (as would doing away with `fs` stuff except for the "indexed" case).

**Comment by Gozala** [comment]
> So this `NotFound` problem reminds me that I made a decision already on this in the HAMT abstraction, doing what you say - mirror what `Map` does and return `undefined` since we don't support storing `undefined`. That should make it feel more natural. So let's just go with a simple `get(cid:CID):Promise<Block|void>`

👍

> is the |void doing the "may be null or undefined here or something else? 

It is bit complicated, if you opt in to `strictNullChecks` (I usually do) in TS it's same as `T|undefined` if you do not then it is `T|undefined|null`. I just like it because it's shorter (and generally don't like differentiating `null` from `undefined` 🥺)

> Should it be |undefined if we want to be strict about not-found returning undefined so there's no additional optionality in what it returns?).

Sure, I don't have have strong opinions either way.

> How about this collection then?
> 
> ```ts
> interface Store {
>   close():Promise<void>
> }
> ```
> 
> ```ts
> interface StoreReader extends Store {
>   has(cid:CID):Promise<boolean>,
>   get(cid:CID):Promise<Block|undefined>
>   getRoots():Promise<CID[]>
> }
> ```
> 
> ```ts
> interface StoreWriter extends Store {
>   put(block:Block):Promise<void>,
>   delete(CID):Promise<boolean>,
>   setRoots([CID]):Promise<void>,
> }
> ```
> 
> ```ts
> interface StoreIterator extends Store {
>   blocks():AsyncIterable<Block>,
>   cids():AsyncIterable<CID>,
>   getRoots():Promise<CID[]>
> }
> ```


I think thing `Store` interface name is a bit confusing and I fell like something like `Resource` would be more adequate. That said I don't really care about naming. So 👍

> The `setRoots()` and `getRoots()` (that last one's duplicated in `StoreReader` and `StoreIterator`, maybe there's an elegant way to factor that out without even more interfaces?) 



I don't find duplication inelegant, in fact I would even duplicate `close` across defined interfaces to git rid of `Store`. Many composition of many interface does create some mental overhead while type checker or a user of API doesn't really care.


> The `setRoots()` and `getRoots()` are interesting here. They're particular to the CAR storage abstraction, but maybe they're not so particular? They're kind of like pins, they're really just graph roots, so maybe they're entirely reasonable to include in generic Store* interfaces and therefore we don't need particular Car* extensions?

I am not sure how general they are. In the context of dag-service they don't seem to make sense, but it could be that I'm missing something ? I think idea of having **index** makes sense, but for the service I think API that allows to add remove roots in one OP would make more sense. E.g. two actors A and B can end up in race condition when they run the code like:

```ts
const addRoots = async (store:StoreReader & StoreWriter, cids:CID[]) => {
  const roots = await store.getRoots()
  await store.setRoots([...roots, ...cids])
}
```

Because order in which things might occur might be:

```
A: store.getRoots()
B: store.getRoots()
A: store.setRoots()
B: store.setRoots()
```

And `B` will undo `A`'s `setRoots`. It also will have no way to guard against that, other than `A` and `B` would have to put store behind some kind of lock. More simply I think it would make more sense for dag-service (at least) to have a different (atomic) API for updating roots e.g.

```ts
interface StoreIndex {
  // Updates roots and resolves promise on completion. Coordinates updates via internal FIFO
  // queue to prevent concurrent updates from interfering with each other.
  updateRoots(update:(roots:CID[]) => CID[]):Promise<void>

  // Alternative optimistic concurrency control (OCC) system is provided to enable updates
  // with no coordination.
  rootsWriter():Promise<OptimisticRootsWriter>
}

interface OptimisticRootsWriter {
   // Sets new roots unless they were updated by another writer in which
   // case errors with `WriterExpired` error.
  set(roots:CID[]):Promise<void>
  // Updates current roots with new roots returned by the `updater` function. Will
  // error with `WriterExpired` if roots were updated by another writer.
  update(updater:(current:T) => T):Promise<void>
}
```

**Comment by Gozala** [comment]
Below is my attempt to capture everything and clean up an API:

```ts

// Performs full parse and keeps everything in memory for you to
// access on demand.
declare class CarReader implements StoreReader, RootsReader {
  static fromBytes(bytes:Uint8Array):CarReader
  static fromFile(fd:FD):CarReader
  static fromIterable(content:AsyncIterable<Uint8Array>|Iterable<Uint8Array>):CarReader

  // It seems to me that `close()` is only relevant for `fd` reader, presumably
  // to close the file descriptor. If so maybe File reader should extracted into
  // it's own thing that just wraps around `CarReader.fromIterable()` and adds
  // close.
  close():Promise<void>
}

// Performs a full parse but only buffers CIDs and the offsets of the relevant
// bytes so reads of individual blocks can seek to that position.
declare class CarIndexedReader implements StoreReader, RootsReader {
  static fromFile(fd:FD): CarIndexedReader

  close():Promise<void>
}

// Mainly good for a non-buffering dump of blocks - turn a stream into a Block
// iterator.
declare class CarBlockIterator implements AsyncIterable<Block>, RootsReader {
  static fromIterable(bytes:AsyncIterable<Uint8Array>):CarBlockIterator
}

// Mainly good for a non-buffering dump of cids - turn a stream into a cids iterator.
declare class CarCIDIterator implements AsyncIterable<CID>, RootsReader {
  static fromIterable(bytes:AsyncIterable<Uint8Array>):CarCIDIterator
}



// Car channel is an async iterable of the car content and a written to write it's content.
declare class CarChannel Channel {
  // Content of this car in form of async iterable. Which will be done (completes for await loop)
  //  when writer is closed.
  content:AsyncIterable<Uint8Array>

  // Represents a writer of this car content.
  writer: CarWriter
}

// A car writer that writes content into a corresponding channel. Writer can't be instantiated
// directly it is exposed through a CarChannel.
declare Class CarWriter implements StoreWriter, RootsWriter {
  close():Promise<void>
}

export {
  CarReader as Reader,
  CarIndexedReader as IndexedReader,
  CarBlockIterator as BlockIterator,
  CarCIDIterator as CIDIterator,
  CarChannel as Channel
}

interface StoreReader {
  has(cid:CID):Promise<boolean>
  get(cid:CID):Promise<Block|undefined>

  blocks() :AsyncIterable<Block>
  cids(): AsyncIterable<CID>
}

interface StoreWriter {
  put(block:Block):Promise<void>,
  delete(cid:CID):Promise<boolean>
}

interface RootsReader {
  getRoots():Promise<CID[]>
}

interface RootsWriter {
  setRoots(roots:CID[]):Promise<void>
}
```

This makes following usage pattern:

```ts
import { Reader, Channel, IndexedReader, CIDIterator, BlockIterator } from "dag-car"

// Implements StoreReader (has, get, cids, blocks),  RootsReader (getRoots) and (close)
const reader = Reader.fromBytes(new Uint8Aray(....))

const file = await fs.open(path)
// Implements StoreReader (has, get, cids, blocks),  RootsReader (getRoots) and (close)
const reader = Car.fromFile(file)

// Implements StoreReader (has, get, cids, blocks),  RootsReader (getRoots) and (close)
// Note that it is different class (an attempts to communicate that it is different from regular buffered Reader) and
// it enables adding things like `IndexedReader.fromBytes` in the future for example.
const indexedReader = IndexedReader.fromFile(file)

// Implements StoreReader (has, get, cids, blocks),  RootsReader (getRoots) and (close)
// Note: We no longer need confusing `complete` suffix here, because it creates a reader
// (not an iterator). This is replacement for `ReadableStream`.
const reader = Reader.fromIterable(fs.createReadStream(path))


// Implements RootsReader (getRoots) and AsyncIterable<Block> so it can be just iterated
// By having `BlockIterator` separate from `CIDIterator` we make error of calling `blocks()` and
// `cids()` impossible to make.
const blocks = BlockIterator.fromItarable((fs.createReadStream(path)))
for await (const block of blocks) {
}

// Implements RootsReader (getRoots) and AsyncIterable<CID> so it can be just iterated
// By having `BlockIterator` separate from `CIDIterator` we make error of calling `blocks()` and
// `cids()` impossible to make.
const cids = CIDIterator.fromItarable((fs.createReadStream(path)))
for await (const cid of cids) {
}


// Instead of taking a `WritableStream` and writing into, we have `Channel` which is like a pipe
// with `content: AsyncIterable<Uint8Array>` and `writer: CatWriter` on the other. Those are
// separate to enable different capabilities on writer and reader end. Writer implements StoreWriter (put, delete),
// RootWriter (setRoots), and (close).
const { content, writer } = new CarChannel()
async reader (content) => {
  for await (const chunk of content) {
     // ...
  }
}


reader(content)
await writer.put(block)
await writer.setRoots([block.cid])
await writer.close()
```

**Comment by Gozala** [comment]
BTW I would further simplify to use polymorphic `.from` 

```ts
declare class CarReader implements StoreReader, RootsReader {
  static from(bytes:Uint8Array):CarReader
  static from(fd:FD):CarReader
  static from(content:AsyncIterable<Uint8Array>|Iterable<Uint8Array>):CarReader

  // It seems to me that `close()` is only relevant for `fd` reader, presumably
  // to close the file descriptor. If so maybe File reader should extracted into
  // it's own thing that just wraps around `CarReader.fromIterable()` and adds
  // close.
  close():Promise<void>
}

// Performs a full parse but only buffers CIDs and the offsets of the relevant
// bytes so reads of individual blocks can seek to that position.
declare class CarIndexedReader implements StoreReader, RootsReader {
  static from(fd:FD): CarIndexedReader

  close():Promise<void>
}

// Mainly good for a non-buffering dump of blocks - turn a stream into a Block
// iterator.
declare class CarBlockIterator implements AsyncIterable<Block>, RootsReader {
  static from(bytes:AsyncIterable<Uint8Array>):CarBlockIterator
}

// Mainly good for a non-buffering dump of cids - turn a stream into a cids iterator.
declare class CarCIDIterator implements AsyncIterable<CID>, RootsReader {
  static from(bytes:AsyncIterable<Uint8Array>):CarCIDIterator
}
```

**Comment by Gozala** [comment]
> it's there because it allowed me to write a more efficient file reader than just using createReadableStream() but it's really not that big a deal and I could offload that burden.

If you want to maintain more efficient file reader optimization, maybe companion library could expose something like `lib.open(path):MoreEfficientFileReader` and `Reader.from(source:Uint8Array|FD|AsyncIterable<Uint8Array>|Iterable<Uint8Array>):CarReader`  so when that extra kick is needed it could be pulled in ?

**Comment by rvagg** [comment]
OK, I like this. Many thanks for your thoughtfulness as always @Gozala.

Channel doesn't quite fit the CAR use-case. I've so far avoided appendable CAR files and even ones you can both read & write and left them as separate operations. Appending would be possible, but not changing the roots since they are at the beginning of the file. It's possible to fake it all just by buffering, I did something like that for datastore-zipcar but I'm not sure it's really a helpful interface, just make a fresh CAR yourself.

But I can see how this _might_ be handy for another type of block store, maybe not a common thing. I'll leave Channel alone for now and see how I go with the rest. Will get some working code and see what it feels like to interact with it this way.

**Comment by Gozala** [comment]
> OK, I like this. Many thanks for your thoughtfulness as always @Gozala.

My pleasure

> Channel doesn't quite fit the CAR use-case. I've so far avoided appendable CAR files and even ones you can both read & write and left them as separate operations. 

I think one of us is misunderstanding something. My assumption was that following

> ```js
> Car.toStream(WritableStream):Car           // supports close(), put(), setRoots()
> ```

Takes a writable stream (**sink**) and returns a `StoreWriter` API to write produced car content into that sink. However, that introduces dependency on nodes `WritableStream` which I think you've expressed interest in removing in the quoted comment:

> Also I think maybe I can factor our ReadableStream and WritableStream and just accept and return AsyncIterator<Uint8Array>, I think that's safe enough to do these days. That would offer quite a bit more flexibility and simplify dependency trees in the browser

I was trying to address that with a `CarChannel` by switching roughly `(Sink) -> Writer` API to `() -> { content: AsyncIterable<Unint8Array>, writer: Writer }`. That way **sink** in form of nodes `WritableStream` is gone instead you have `content: AsyncIterable<Uint8Array>` that user can pipe into node `WritableStream` like:

```ts
const { content, writer } = new Channel()
const writable = fs.createWriteStream(path)
stream.pipeline(content, writable)

// do actual encoding
await writer.put(block)
// ...
await writer.close()
```

In browser contexts `content: AsyncIterable<Uint8Array>` be piped just as easily. 


> Appending would be possible, but not changing the roots since they are at the beginning of the file. It's possible to fake it all just by buffering, I did something like that for datastore-zipcar but I'm not sure it's really a helpful interface, just make a fresh CAR yourself.

Before I dive into this further, I want to call out that `CarChannel` makes no assumptions about that. If roots are written in the beginning then I understand how that would make it impossible to stream any content until writer closes. This also suggests that:

1. `setRoots` can be called only once.
2. Car content will need to buffer until `setRoots` is called.

It would be nice to capture invariant no 1 so it is impossible to do so. Maybe something like:

```ts
declare class CarChannel {
  // Constructor takes root blocks because:
  // 1. Roots can be set only once for car
  // 2. Blocks instead of CIDs so that it's impossible to not put those blocks.
  // 3. So that no `put`s need to get buffered until roots are set and content can start streaming right away.
  constructor(roots:Block[]):CarChannel

  content: AsyncIterable<Uint8Array>
  writer: CarWriter
}

declare Class CarWriter implements StoreWriter {
  close():Promise<void>
}

interface StoreWriter {
  put(block:Block):Promise<void>,
  delete(cid:CID):Promise<boolean>
}
```

However how can `delete` be a thing though unless things are buffered ? Would not put block be already be written into car content so that delete would be impossible ?

**Comment by Gozala** [comment]
> However how can `delete` be a thing though unless things are buffered ? Would not put block be already be written into car content so that delete would be impossible ?

Ah I see now that I overlooked that following had no `delete`

> ```ts
> Car.toStream(WritableStream):Car           // supports close(), put(), setRoots()
> ```

So it all makes sense.

I think you have two kind of readers one that buffers and other that doesn't. And only now I'm realizing that you have a writer that doesn't buffers while I was mistakenly thinking it was.

**Comment by rvagg** [comment]
Yeah, `delete()` was a holdover from datastore that can go away, but I can imagine it coming back if/when there's a CAR interface that's mutable and it just does a full rewrite, I'm just not very interested in doing that.

And yes, I can see what you're trying to achieve with the Channel now, it's really just a converter, put stuff and Uint8Arrays get emitted on the async iterator. Do we have prior art for that kind of pattern out there in non-Node land? Having a clear `WritableStream` that you can accept is nice in enforcing a pattern and it makes documentation easier. The unfortunate side effect of all of these APIs is that they're going to require a lot more explanation except where we can lean on existing patterns.

**Comment by Gozala** [comment]
> Do we have prior art for that kind of pattern out there in non-Node land?

If you mean generally [go](https://gobyexample.com/channels), [rust](https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html), [clojure](https://clojuredocs.org/clojure.core.async/chan) all use channels, although to my knowledge Rust is only one that separates reader and writer ends. 

If you meant in JS, mozilla code base uses [`nsIPipe`](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIPipe)s all over the place and there is also [MessageChannel](https://developer.mozilla.org/en-US/docs/Web/API/MessageChannel) API on the web. I think pull-streams also embraced concept of no writable streams but sources and sinks.

> Having a clear WritableStream that you can accept is nice in enforcing a pattern and it makes documentation easier. 

There's a web [`WritableStream`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStream)s but unfortunately they are not well supported, so in most cases people use [`ReadableStream`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream)s instead.

I would call out one nice (but probably impractical) property of channel based approach:

```ts
const { content, writer } = new Channel()
// You can create car reader from the output of the writer, without
// having to introduce duplex streams and all.
const reader = Reader.from(content)
```

**Comment by rvagg** [comment]
Back to `CarChannel`, what if the `StoreWriter` was itself an async iterable?

```ts
interface StoreWriter {
  put(block:Block):Promise<void>,
  delete(cid:CID):Promise<boolean>
  [Symbol.asyncIterator](): AsyncIterator<Uint8Array>
}
```

Would that work? Then I could:

```js
import Readable from 'stream'
import { CarWriter } from 'datastore-car'

const writer = new CarWriter(/* maybe roots can go in here? */)
Readable.from(writer).pipe(fs.createWriteStream('/path/to/file.car')) // pipeline() would probably be nicer
await writer.put(Block.create(...))
await writer.put(Block.create(...))
await writer.put(Block.create(...))
await writer.close()
```

Am I overlooking something or is that workable? It would remove the need for an unnatural-feeling API (yes Channels are common in other languages, but it's not a native JavaScript-feeling API so it'd be nice to avoid it if possible).

**Comment by rvagg** [comment]
Implemented this:

```js
      const writer = CarWriter(Block).create(rootsArray)
      Readable.from(writer).pipe(/* some place */)
      for (const block of listOfBlocks) {
        await writer.put(block)
      }
      await writer.close()
```

**Comment by rvagg** [comment]
The tricky thing about an API like this is what to do about back-pressure. Currently I have it queuing bytes and it'll all spew forth when you pipe it, but the ideal is that `await writer.put()` will block to allow back-pressure. In that case you need to have already sent your output somewhere. This is not unreasonable, it'd be bad practice to queue it all up in memory, but perhaps you have a small amount of data to process and can afford the memory, but when back-pressure is in place `await writer.put()` may end up locking up your process unless you're already draining the writer out the other end. At least with the old streams API you had to create the write stream with a target, so you're forced to order the operations properly. Just shifting the `Readable.from(writer).pipe()` down below the `for` loop is going to cause headaches one way or another.

**Comment by Gozala** [comment]
> Back to `CarChannel`, what if the `StoreWriter` was itself an async iterable?

Sorry just catching up on this. Originally I was going to suggest that, but then reconsidered, because:

It hosts two capabilities into one. E.g. Actor that is able to read gets capability to write as well. In practice that tends to be a bad choice because there are cases where reader and writer must be different actors and inability to decouple those capabilities make it impossible. I personally would choose separation of concerns / capabilities over API convenience. And when convenience is really desired  I tend to go with a model where channel is implements both reader and writer APIs by delegating to own reader and writer properties. That way combined capabilities can be shared via same object, but sharing individual capabilities is also possible.

**Comment by Gozala** [comment]
> but the ideal is that await writer.put() will block to allow back-pressure.

That was intention behind `put(...):Promse<void>`. Well in ideal world we would go full circle to [CSP](https://en.wikipedia.org/wiki/Communicating_sequential_processes) and have an abstract buffer concept that you use when creating a channel which can block once it's filled. Which is what go and clojure channels do, and to some degree `WritableStream`'s queuing strategy attempts to do, but I really wish whatwg group considered provided feedback more closely 🤷‍♂️, oh well.

**Comment by rvagg** [comment]
OK, well for now I've just gone with the "writer is an iterator" approach and have implemented backpressure. This is all on the rvagg/storage-iface branch, see tests exercising the API @ https://github.com/ipld/js-datastore-car/blob/rvagg/storage-iface/test/test-writer.js#L96-L107, which seems OK to me. There's some awkwardness in having to ensure you're sending the output somewhere before you start awaiting on input otherwise you could get yourself stuck, but that should be easily discoverable when consuming the API.

I have a bunch more refactoring to do to make the internal organisation make more sense but I'll revisit your concern about dual capabilities when the innards feel cleaner. I might come back to you to help outline some cases where this is genuinely a problem because I don't find the Channel abstraction very JS-like (although so much of these new Promise based APIs don't feel JS-like, so maybe it's not that big a leap!).

I'm thinking that I might deprecate this library and publish the new one as `@ipld/car` to ditch the "datastore" name. Will figure that out this week hopefully.

**Comment by mikeal** [comment]
Love the “writer is an iterator.”

We should definitely publish to `@ipld/car`

**Comment by Gozala** [comment]
> I might come back to you to help outline some cases where this is genuinely a problem because I don't find the Channel abstraction very JS-like

I don't mean to press my point here, but since you intend to revisit this & I'm in the mood to elaborate my point, I'll leave this comment here for the future:

- I understand your point about non JS-like, but then again as you pointed out Promises were not very JS like until they have become and so were other patterns. In fact, immutable data structures isn't very JS like either, but it is our bread and butter :P That is to suggest that JS-likeness probably should fall behind other API design concerns.
- My capabilities argument is that you want to encapsulate write capability `W` seperate from read capability `R`. As this would enable passing `W` capability to an actor in the system enabling it  write but not read. Likewise passing `R` capability to an actor will enable it to read, but not write. If both capabilities were  encapsulate togather `RW` it becomes impossible to have an actors in the system that can only read but not write, or write but not read because passing `RW` would allow them to do both.

   Fairly often readers and writers tend to be different actors in the system. Sometimes you can have multiple readers but single writer (I understand that does not apply here).

   Analogy with file descriptors may make sense here, we tend to use open them in read modes for reading and write mode mode for writing. 

   Another analogy I could make is asymmetric encryption, if I want to give a person (cap)ability to read my (encrypted) messages, I'll share a public key with them, not the private key because later would allow you them to impersonate me.

   For what it's worth [WASI is build around](https://hacks.mozilla.org/2019/03/standardizing-wasi-a-webassembly-system-interface/) [Object Capabilities Model](https://en.wikipedia.org/wiki/Object-capability_model).

- Separation of concerns is a different spin on the same core idea (where capabilities comes security perspective and SoC is from modularity perspective). While it is tempting to hang async iteration on writer, it introduces ambiguity that needs to be communicated with consumers downstream (via documentation or something). Here is the sample:
   1. When an actor receives a `StoreWriter` is it ok for it to call `put` or is actor supposed to only read ?
   2. When an actor is given `StoreWriter` , is it ok for it to iterate over the content or is it supposed to only write ?
   3. This list continues to every single method. While answers may seem obvious here, in my experience if there is the thing to call someone will for some unanticipated reasons and cause unexpected problems.
  
   By separating reader and writer ends, above question disappear. There is an ambiguity in intent to be communicated (via docs or other means) because if actor gets a thing it can only read from that is only thing it can do. Better yet, if there is an actor that needs to both read and write, it becomes obvious from just the surface as it will be be passed both read and write ends. If read and write end are combined, it becomes a lot more difficult to identify who does the reading and who does the writing and we can just hope for good naming conventions or comments.

- Separation of concerns also tends to have nice long term effects. E.g. if your reader part is just an `AsyncIterable<Uint8Array>` revisiting writer API will not break any reader code. However if you combine those together any change is a breaking change because all the code downstream also tends to not separate read / write concerns clearly.

- It is hard to provide a very specific example, but here is my attempt given recent endeavor. With a shared worker for IPFS I can imagine wanting to have car writer in the main thread and reader in the worker thread. It is certainly possible to pipe output of `StoreWriter` over the message port to the worker, but then again if you have `StoreWriter` abstraction chances are that things build around it would expect `StoreWriter` and not just `AsyncIterable<Uint8Array>`. This may seem very hypothetical, but I've encountered quite a few assumptions like that during mentioned endeavor.

  It does make sense if you think about it. With a single thread there was no considerations to be made about what piece of code should run where there was no split to be made. However when separation of concerns is a designed concern, it leads to separate actors dealing with separate concerns and there for are easier to move around (e.g across different threads).

**Comment by rvagg** [comment]
This work moved over to https://github.com/ipld/js-car which is ready to be published now. The API mostly matches the discussion here. I even (reluctantly) gave in to the "channel" language after trying hard to squeeze in other terms, it ended up being the most logical for the pattern.

Typed API is outlined @ https://github.com/ipld/js-car/blob/master/api.ts
